### ISSUE TITLE
High system load following 2 recent Docker upgrades

### ISSUE THREAD

**Issue by toby-griffiths** [moby/moby#47214]
### Description

We've experienced really high CPU load on our server twice recently, both times immediatly at the point that `yum` starts the upgrade of Docker packages.

We are running on CentOS

### Reproduce

1. Install Plesk 18.0.58;
2. Install Docker via Plesk;
3. Configure automatic upgrades;

I suspect there may be a 'start these Docker services' step also that might make a difference, but I've not Idea what is causing the high load.

### Expected behavior

Docker should be upgraded and conters restart without any issues, and we should only see brief CPU usge spikes as the containers start up.

### docker version

```bash
After last night's upgrade…

Client: Docker Engine - Community
 Version:           25.0.1
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        29cf629
 Built:             Tue Jan 23 23:12:51 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          25.0.1
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       71fa3ab
  Built:            Tue Jan 23 23:11:50 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.27
  GitCommit:        a1496014c916f9e62104b33d1bb5bd03b0858e59
 runc:
  Version:          1.1.11
  GitCommit:        v1.1.11-0-g4bccb38
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    25.0.1
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.2
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 193
  Running: 70
  Paused: 0
  Stopped: 123
 Images: 503
 Server Version: 25.0.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: loki
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
Swarm: active
  NodeID: l603mylmj6hnh95at639cy4w0
  Is Manager: true
  ClusterID: vsayqmcuktjv67jgmtgjqm5kk
  Managers: 1
  Nodes: 1
  Default Address Pool: 10.0.0.0/8
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 31.193.2.114
  Manager Addresses:
   31.193.2.114:2377
 Runtimes: runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: a1496014c916f9e62104b33d1bb5bd03b0858e59
 runc version: v1.1.11-0-g4bccb38
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
 Kernel Version: 3.10.0-1160.95.1.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 25.34GiB
 Name: id79706
 ID: THRG:QZBK:5427:DLKF:M4HF:Y6TQ:32CN:PTR3:IWUG:HOEC:C5T7:TTNI
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

Here is our System Load graph with the upgrade times marked.  This shows last nights issue as more severe.

The initial upgrade issue was causing quite a bit of issues running things like `ps` and `htop`, but fortunately the web server was still responding OK at that point.  The cliff edge marked in blue was a server reboot we performed to get things under control, and even that took far longer than typical to come back up again, and I had to contact our support team to get it back online.

![image](https://github.com/moby/moby/assets/4817007/ef24d548-8fc7-457d-9ca1-7a16cc3a6d78)

System info…
CentOS Linux release 7.9.2009 (Core)

Plesk…
18.0.58.2
Plesk Obsidian 18.0

The 2 recent yum upgrades…

```
Loaded plugins: fastestmirror
Transaction ID : 2367
Begin time     : Thu Jan 25 03:25:21 2024
Begin rpmdb    : 1258:c9f9a40e049bf4d96c3bf8f46d1daabc1cc1f718
End time       :            03:30:30 2024 (5 minutes)
End rpmdb      : 1258:94fbfead98d6a40bb9bd43a96bdee5b52ce5fca5
User           : System <unset>
Return-Code    : Success
Transaction performed with:
    Installed     rpm-4.11.3-48.el7_9.x86_64                      @updates
    Installed     yum-3.4.3-168.el7.centos.noarch                 @base
    Installed     yum-metadata-parser-1.1.4-10.el7.x86_64         @anaconda
    Installed     yum-plugin-fastestmirror-1.1.31-54.el7_8.noarch @updates
Packages Altered:
    Updated docker-ce-3:25.0.0-1.el7.x86_64               @plesk-ext-docker
    Update            3:25.0.1-1.el7.x86_64               @plesk-ext-docker
    Updated docker-ce-cli-1:25.0.0-1.el7.x86_64           @plesk-ext-docker
    Update                1:25.0.1-1.el7.x86_64           @plesk-ext-docker
    Updated docker-ce-rootless-extras-25.0.0-1.el7.x86_64 @plesk-ext-docker
    Update                            25.0.1-1.el7.x86_64 @plesk-ext-docker
    Updated docker-compose-plugin-2.24.1-1.el7.x86_64     @plesk-ext-docker
    Update                        2.24.2-1.el7.x86_64     @plesk-ext-docker
history info
```

```
Loaded plugins: fastestmirror
Transaction ID : 2362
Begin time     : Sat Jan 20 03:13:20 2024
Begin rpmdb    : 1258:aa696fc965824607ba376bcba2c42cedf14780cc
End time       :            03:17:44 2024 (264 seconds)
End rpmdb      : 1258:38f5b6c8b8b47fbdb03d19a8d62799c88e36c680
User           : System <unset>
Return-Code    : Success
Transaction performed with:
    Installed     rpm-4.11.3-48.el7_9.x86_64                      @updates
    Installed     yum-3.4.3-168.el7.centos.noarch                 @base
    Installed     yum-metadata-parser-1.1.4-10.el7.x86_64         @anaconda
    Installed     yum-plugin-fastestmirror-1.1.31-54.el7_8.noarch @updates
Packages Altered:
    Updated docker-buildx-plugin-0.11.2-1.el7.x86_64      @plesk-ext-docker
    Update                       0.12.1-1.el7.x86_64      @plesk-ext-docker
    Updated docker-ce-3:24.0.7-1.el7.x86_64               @plesk-ext-docker
    Update            3:25.0.0-1.el7.x86_64               @plesk-ext-docker
    Updated docker-ce-cli-1:24.0.7-1.el7.x86_64           @plesk-ext-docker
    Update                1:25.0.0-1.el7.x86_64           @plesk-ext-docker
    Updated docker-ce-rootless-extras-24.0.7-1.el7.x86_64 @plesk-ext-docker
    Update                            25.0.0-1.el7.x86_64 @plesk-ext-docker
    Updated docker-compose-plugin-2.21.0-1.el7.x86_64     @plesk-ext-docker
    Update                        2.24.1-1.el7.x86_64     @plesk-ext-docker
history info
```

Happy to provide more info if you let me know what you need.

**Comment by toby-griffiths** [comment]
It's probable worth mentioning that I shut down some non-essential Docker services following the initial issue on Saturday morning (between the first Docker upgrade and my implementing resource limits), and these were all redeployed when the resource limits were added that evening, so someting there could possibly be causing the initial ramp, perhaps.

You can also see that the System Load (& also the CPU, not on the graph) remained low following the reboot, but all the Docker services were definitelty running during most of this time.

Any suggestions on how we can determine the cause of the System Load would be appreciated as well.

**Comment by toby-griffiths** [comment]
I couldn't check at the time.  `htop` was just showing a blank screen and hanging, and this time around I couldn't ssh in at all.

Are there other tools to view the CPU load, other than htop?

**Comment by S-S-X** [comment]
You could try to have high priority shell waiting in background or use some simple high priority snmp exporters for system stats or even start simple script (again high priority) to start collecting data on host to say just some local file you can retrieve later.

One of these usually works with stuff that might lock you out and are rather simple to set up. Well, bit more complicated than just running a command but there's a lot of info available.

Basically straightforward workaround to get some CPU time is to simply adjust priority of your/other processes.

**Comment by thaJeztah** [comment]
Closing, as this went stale and does not have enough information to make it actionable.

CentOS 7 also reached EOL, and we no longer build packages for it.
